### ISSUE TITLE
Medic call might not work with 60 fps

### ISSUE THREAD

**Issue by arctic5** [Gang-Garrison-2/Gang-Garrison-2#155]
similar to https://github.com/Gang-Garrison-2/Gang-Garrison-2/issues/25

**Comment by harmjanhaisma** [comment]
Confirmed.

**Comment by Medo42** [comment]
Reproduced locally with 30fps server and 60fps client. It seems to be roughly a 50/50 chance of the medic call working or not. I don't quite understand the issue yet since the keystate system is actually quite good about remembering press/release events until they are processed (in `Character.BeginStep`), but it seems like something that should be easy to fix.

**Comment by Medo42** [comment]
The issue is that we only send the medic (and also the taunt) key bit as `1` in the keybyte when it was pressed *just now*. This runs counter to what the keybyte should represent - the current *state* of the keys. Whether they were just pressed or not is already handled later, and quite robustly, in the Character code.
Since a 60fps client will only actually send its keybyte every second step, there is a 50% chance that the medic input will be missed, since it only lasts for one step regardless of how long you press the button.
A fix was added for the taunt button already by ensuring that a keybyte is *always* sent when the user just pressed the taunt button. However, this is still only a partial fix. The *server* will now always register a taunt button press, since it will see every keybyte update pushed by the client. However, if the networking situation conspires such that the server will receive two keybyte updates from a client in the same step, and the first one of those is the one which has the taunt bit set to 1, then the server will never push a keybyte with set taunt bit to any client. Worse, since taunting is not sync'd (it should be, really), this means you now get a situation where you are taunting on the server but not on any client, causing desync teleporting and confusion until the taunt ends.
Even under perfect networking and timing conditions, this is actually quite likely to happen with Wareya's taunt fix given a 60fps client and 30fps server, because the taunt keybyte update can be sent between two "normal" updates happening at the usual 30fps rate.

Ok, long analysis here. Now on to conclusions: IMO the issue is misuse of the keybyte. Since we're only synchronizing the current state to everyone and then calculate the press/release info from there, and since client updates can arrive at irregular intervals but are pushed out in collected form, we really shouldn't have single-update events on there. Either we treat the medic key like the other keys and send its bit as 1 as long as it's held down (pressed-just-now-check done by the keybyte receiver), or we send a separate event and remove the medic key from the keybyte (pressed-just-now-check done by the sending client).
I'd just go for the former, except that the only thing the medic key does is already covered by the chatbubble event, so I'll just fix it to use that.

tl;dr: I just wrote way too much text about what will be a really small fix :)
